### PR TITLE
[5.3][Diagnostics] Tailored diagnostics for missing member failure when in…

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
… literal collection context

cc @xedin 

Resolves SR-12672.
Resolves: rdar://problem/62383753

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@LucianoPAlmeida There is no need for `dyn_cast` on the anchor, it's still just an `Expr *`.

**Comment by LucianoPAlmeida** [comment]
> @LucianoPAlmeida There is no need for `dyn_cast` on the anchor, it's still just an `Expr *`.

Ah, I adjust real quick :) I wasn't sure so I looked at other references in the file the ones I've seen were like this

**Comment by LucianoPAlmeida** [comment]
The Windows build failed with no identified reason ... should we kick it again?

**Comment by xedin** [comment]
Let's get that `get<>` removed and I'm going to kick it again.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Review by devincoughlin — APPROVED** [review]

**Comment by LucianoPAlmeida** [comment]
The error on Windows `CSDiagnostics.cpp(3432): error C3535: cannot deduce type for 'auto *' from 'swift::constraints::TypedNode'` ... so I think we should put the `get<>` back?

**Comment by xedin** [comment]
Hm, let me check what is going on here...

**Comment by LucianoPAlmeida** [comment]
On linux `CSDiagnostics.cpp:3432:9: error: variable 'expr' with type 'auto *' has incompatible initializer of type 'TypedNode' (aka 'PointerUnion<const swift::Expr *, const swift::TypeLoc *, const swift::VarDecl *, const swift::Pattern *>`

So, I just put the `get<>`back and replace the `cs.getParentExpr` with `findParentExpr` from `FailureDiagnostics`

**Comment by xedin** [comment]
Sorry, it's my bad - use `castToExpr`, looks like only diagnostic refactoring got into 5.3 branch :)

File: lib/Sema/CSDiagnostics.cpp [hunk 1]
```diff
@@ -3424,6 +3427,37 @@ bool MissingMemberFailure::diagnoseForDynamicCallable() const {
...
+  if (!(parentExpr && isa<UnresolvedMemberExpr>(expr)))
+    return false;
+
+  auto parentType = getType(parentExpr);
+
+  if (!cs.isCollectionType(parentType) && !parentType->is<TupleType>())
+    return false;
+
+  if (isa<TupleExpr>(parentExpr)) {
+    parentExpr = cs.getParentExpr(parentExpr);
```
Comment (xedin) [inline lib/Sema/CSDiagnostics.cpp:3445]: I think this has to be `findParentExpr` as well :/

Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: Fixed!

**Comment by xedin** [comment]
Ok, let's do this!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by LucianoPAlmeida** [comment]
Any reason for macos and windows not started?

**Comment by xedin** [comment]
Most likely they are in the queue, let’s give them some time :)

**Comment by LucianoPAlmeida** [comment]
Ok I just made sure it builds locally! So I think is fine now :) 
@xedin mind kick the tests again?

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@LucianoPAlmeida Thank you for fixing this!

**Pull request merged**